service/worker/scheduler/calendar.go

```diff
@@ -0,0 +1,385 @@
+			}
+		}
+		if h >= 24 {
+			d, h = d+1, 0
+		}
+		if d > daysInMonth(mo, y) {
+			mo, d = mo+1, 1
+		}
+		if mo > time.December {
+			y, mo = y+1, time.January
+		}
+		if y > maxCalendarYear {
+			break Outer
+		}
+		// try to match year, month, etc. from outside in
+		if !cc.year(y) {
+			y, mo, d, h, m, s = y+1, time.January, 1, 0, 0, 0
+			dstoffset = 0
+			continue Outer
+		}
+		for !cc.month(int(mo)) {
+			mo, d, h, m, s = mo+1, 1, 0, 0, 0
+			dstoffset = 0
+			if mo > time.December {
+				continue Outer
+			}
+		}
+		for !cc.dayOfMonth(d) || !cc.dayOfWeek(int(time.Date(y, mo, d, h, m, s, 0, cc.tz).Weekday())) {
+			d, h, m, s = d+1, 0, 0, 0
+			dstoffset = 0
+			if d > daysInMonth(mo, y) {
+				continue Outer
+			}
+		}
+		for !cc.hour(h) {
+			h, m, s = h+1, 0, 0
+			dstoffset = 0
+			if h >= 24 {
+				continue Outer
+			}
+		}
+		for !cc.minute(m) {
+			m, s = m+1, 0
+			if m >= 60 {
+				continue Outer
+			}
+		}
+		for !cc.second(s) {
+			s = s + 1
+			if s >= 60 {
+				continue Outer
+			}
+		}
+		// everything matches
+		nextTs := time.Date(y, mo, d, h, m, s, 0, cc.tz)
+		// we might have reached a nonexistent time that got jumped over by a dst transition.
+		// we can tell if the hour is different from what we think it should be.
+		if nextTs.Hour() != h {
+			h, m, s = h+1, 0, 0
+			continue Outer
+		}
+		return nextTs.Add(dstoffset)
+	}
+
+	// no more matching times (up to max we checked)
+	return time.Time{}
+}
+
+// Returns a function that matches the given integer range/skip spec.
+// The function _may_ return true for values that are out of range.
+func makeMatcher(s, def string, min, max int, parseMode parseMode) (func(int) bool, error) {
+	s = strings.TrimSpace(s)
+	if s == "" {
+		s = def
+	}
+	// easy special cases
+	if s == "*" {
+		return func(int) bool { return true }, nil
+	} else if s == "0" {
+		return func(v int) bool { return v == 0 }, nil
+	}
+	if max <= 63 {
+		return makeBitMatcher(s, min, max, parseMode)
+	} else if max < math.MaxInt16 {
+		return makeSliceMatcher(s, min, max, parseMode)
+	}
+	return nil, errOutOfRange
+}
+
+func makeBitMatcher(s string, min, max int, parseMode parseMode) (func(int) bool, error) {
+	var bits uint64
+	add := func(i int) { bits |= 1 << i }
+	if err := parseStringSpec(s, min, max, parseMode, add); err != nil {
+		return nil, err
+	}
+	if parseMode == parseModeDow {
+		bits |= bits >> 7 // allow 7 or 0 for sunday
+	}
+	return func(v int) bool { return (1<<v)&bits != 0 }, nil
+}
+
+func makeSliceMatcher(s string, min, max int, parseMode parseMode) (func(int) bool, error) {
+	var values []int16
+	add := func(i int) { values = append(values, int16(i)) }
+	if err := parseStringSpec(s, min, max, parseMode, add); err != nil {
+		return nil, err
+	}
+	return func(v int) bool {
+		for _, value := range values {
+			if int(value) == v {
+				return true
+			}
+		}
+		return false
+	}, nil
+}
+
+func parseStringSpec(s string, min, max int, parseMode parseMode, f func(int)) error {
```

Comment:
need comments on what are the expected specs.
would be good to put some input example to help reader understand this code